enhancements/olm/platform-operators.md

```diff
@@ -0,0 +1,609 @@
+
+### Implementation Details
+
+> Note: The following sub-headers may frequently refer to OLM 1.x terminology. It's recommended to read through the ["Common Terminology"](#Common-Terminology) header for those less familiar with recent OLM developments.
+
+#### The POM Component
+
+The PO manager would abstract the process of sourcing registry+v1 bundles from catalog(s), choosing the "best bundle" to install for each desired package, and ensuring those packages get installed on the cluster.
+
+In order to determine the set of bundles that need to be installed, the PO manager can interact with the default, `redhat-operators` CatalogSource, and use OLM's existing registry gRPC APIs to list the available bundles from those catalog contents.
+
+> The OLM marketplace component is now an optional CVO component. Admins can disable this component using the component capabilities functionality introduced in previous 4.x releases. Available workarounds include admins creating their own CatalogSource resources.
+> This behavior may require explicit documentation, or expanding the phase 0 design scope to be able to accommodate all the catalog sources in the cluster.
+
+After determining the set of bundles that need to exist in order to satisfy admin package intent, the PO manager's installation layer will need to attempt to install those bundles. The management of those bundle contents will be delegated to OLM's rukpak layer, and the POM component would be responsible for generating an underlying BundleDeployment resource for the bundle that was selected.
+This underlying, cluster-scoped BundleDeployment resource will contain owner references to the parent PlatformOperator custom resource the manager is reconciling.
+
+The POM component will also be responsible for bubbling up any failure states during reconciliation:
+
+- Failure to successfully source a bundle from the admin configured OLM package name.
+- Failure to successfully install the OLM-based operator after determining a bundle that meets the desired package name intent.
+
+In both cases, the POM component will relay those failure states back to the PlatformOperator custom resource's status sub-resource that it's currently reconciling. Any PlatformOperator custom resources that are expressing failing status condition types will eventually be reflected in the "aggregate" ClusterOperator resource.
+
+Admins may elect to install or delete a PO package after cluster creation. This process would include creating, or deleting, a PlatformOperator custom resource. In order to delete a PO after cluster creation, admins would need to delete the top-level PlatformOperator custom resource for that package.
+
+**Note**: Deleting the top-level PlatformOperator custom resource will result in the cascading deletion of bundle contents. This may result in Kubernetes garbage collecting cluster-scoped resources, e.g. CRDs, which can impact workload health in unsupported configurations.
```

Comment:
Why does this differ? What's the benefit here? Why aren't we protecting from removing CRDs as well to avoid data loss?

Reply:
I'm not sure what @joelanford and @timflannagan have discussed, but from my perspective the current olm behavior of not deleting operand content when the operator is uninstalled is not behavior we want to carry forward.  I'm not sure what the plan is for olm 1.0 behavior generally(though i think it is also to cleanup operands on uninstall), but for platform operators at least we have an opportunity here to declare a new flow.

In conjunction with that, of course, is that we need to give admins the tools/power to "fix" broken installed operators without uninstalling them.  (Today OLM itself relies heavily on "uninstall/reinstall" as a valid flow to fix broken things, so deleting operands during that flow would be a problem).

Reply:
Yup, I believe that full removal of operator contents are one of the functional requirements for OLM 1.x that we have planned. That said, I think this current implementation where we simply just allow cascading deletion will likely be improved on over time, and we'll be "smarter" around CRD deletion. See https://github.com/operator-framework/rukpak/issues/454 for an example of situations we'd like to be able to accommodate in the future.

Reply:
Ok, I guess we can table the data loss concerns for the OLM 1.x era, otherwise this sounds ok so long as it's documented